### PR TITLE
[JENKINS-24097] Allow configurable colors (for colorblind accessibility)

### PR DESCRIPTION
**Pull request by gdelory**
We have a large community of users using our servers and some of them are colorblind. The warning color is impossible to read (even for not colorblind people IMO), so I think that should be configurable. Settings have been added to the advanced section and current colors have been kept as default.

### CODE REVIEW

**Comment by slide** [comment]
@jenkinsci/code-reviewers Please review changes.

File: src/main/java/hudson/plugins/logparser/LogParserDisplayConsts.java [hunk 1]
```diff
@@ -11,9 +13,13 @@
...
     public LogParserDisplayConsts() {
         // Color of each status
-        colorTable.put(LogParserConsts.ERROR, "red");
-        colorTable.put(LogParserConsts.WARNING, "orange");
-        colorTable.put(LogParserConsts.INFO, "blue");
+        LogParserPublisher.DescriptorImpl descriptor = (LogParserPublisher.DescriptorImpl) Jenkins.getInstance().getDescriptor(LogParserPublisher.class);
+        String colorError = descriptor.getColorError();
+        String colorWarning = descriptor.getColorWarning();
+        String colorInfo = descriptor.getColorInfo();
+        colorTable.put(LogParserConsts.ERROR, colorError == null || colorError.isEmpty() ? "red" : colorError);
```
Comment (slide) [inline src/main/java/hudson/plugins/logparser/LogParserDisplayConsts.java]: I would think the check for null or isEmpty should be done in the descriptor method so you don't have to check it everywhere else. Check in the getColorError for null or isEmpty and set a sane default ("red' for example). Then you just call the method everywhere and it handles it for you.

File: src/main/java/hudson/plugins/logparser/LogParserPublisher.java [hunk 2]
```diff
@@ -206,6 +221,9 @@ public boolean configure(final StaplerRequest req, final JSONObject json)
                     "log-parser.").toArray(new ParserRuleFile[0]);
             useLegacyFormatting = json.getJSONObject("log-parser").getBoolean(
                     "useLegacyFormatting");
+            colorWarning = json.getJSONObject("log-parser").getString("colorWarning");
```
Comment (slide) [inline src/main/java/hudson/plugins/logparser/LogParserPublisher.java]: Is any data validation needed for these? What if they are empty?

File: src/main/java/hudson/plugins/logparser/LogParserPublisher.java [hunk 3]
```diff
@@ -169,6 +169,9 @@ public void perform(Run<?, ?> build, FilePath workspace, Launcher launcher, Task
         public static final DescriptorImpl DESCRIPTOR = new DescriptorImpl();
         private volatile ParserRuleFile[] parsingRulesGlobal = new ParserRuleFile[0];
         private boolean useLegacyFormatting = false;
+        private String colorWarning = "orange";
```
Comment (slide) [inline src/main/java/hudson/plugins/logparser/LogParserPublisher.java]: Should these default to the descriptor (global) color selection?

File: src/main/java/hudson/plugins/logparser/LogParserDisplayConsts.java (hunk 1 above)
Comment (gdelory) [inline src/main/java/hudson/plugins/logparser/LogParserDisplayConsts.java]: Thanks for the ultra quick feedback, agreed on all your comments :) will make the changes.

File: src/main/java/hudson/plugins/logparser/LogParserPublisher.java (hunk 3 above)
Comment (gdelory) [inline src/main/java/hudson/plugins/logparser/LogParserPublisher.java]: Sorry I don't get this one, what are you calling *global* color selection? It's already in the descriptor class, I must be missing something here.

Comment (slide) [inline src/main/java/hudson/plugins/logparser/LogParserPublisher.java]: You're right, I was looking at the filename only and thinking the colors were in the publisher and not the descriptor.

**Comment by gdelory** [comment]
I've made the requested changes. It looks like it caused the build to fail, probably because of the failure in the test. The thing is that it has nothing to do with my commit, I think the existing code is failing (it might be too old). What does it mean, does it need to be fixed anyway before being merged?

Sorry about that, not sure why it worked before those changes and not now. It might be because I'm member of the repo since then (for another plugin)?

File: src/main/java/hudson/plugins/logparser/LogParserPublisher.java [hunk 4]
```diff
@@ -203,15 +213,15 @@ public boolean getLegacyFormatting() {
         }
         
         public String getColorError() {
-            return colorError;
+            return colorError == null || colorError.isEmpty() ? defaultErrorColor : colorError;
```
Comment (slide) [inline src/main/java/hudson/plugins/logparser/LogParserPublisher.java]: You could just check it once and then set the colorError/colorWarning/colorInfo to the defaultErrorColor, etc.

File: src/main/java/hudson/plugins/logparser/LogParserPublisher.java [hunk 5]
```diff
@@ -206,9 +240,36 @@ public boolean configure(final StaplerRequest req, final JSONObject json)
                     "log-parser.").toArray(new ParserRuleFile[0]);
             useLegacyFormatting = json.getJSONObject("log-parser").getBoolean(
                     "useLegacyFormatting");
+            JSONObject colors = json.optJSONObject("log-parser");
+            if (colors != null) {
+                colorWarning = colors.optString("colorWarning");
```
Comment (daniel-beck) [inline src/main/java/hudson/plugins/logparser/LogParserPublisher.java]: Is it deliberately possible to define invalid colors? Shouldn't these assignments conditional on being valid colors?

Comment (gdelory) [inline src/main/java/hudson/plugins/logparser/LogParserPublisher.java]: What do you mean? If a user is specifying a string which is not a color, like "blah"? The thing is how do you check the string is a valid css color without a giant array of all the valid colors? I'm already doing form validation on hex color and string (which should be letters only words), but I don't know how to test that.

Comment (slide) [inline src/main/java/hudson/plugins/logparser/LogParserPublisher.java]: Will anything blow up for the user if they specify something that is incorrect? If so, then that needs to be handled somehow.

Comment (gdelory) [inline src/main/java/hudson/plugins/logparser/LogParserPublisher.java]: No it won't, basically the plugin creates a span with a style color: thecolor, so it won't change the color that's it.

**Comment by nfalco79** [comment]
I would suggest to add [JENKINS-24097] prefix to all commits (you can also squash) to bind those commit to issue.

**Comment by oleg-nenashev** [comment]
Or the commits can be just squashed

**Comment by gdelory** [comment]
I think I managed this without breaking everything ;) If this PR looks good to you, someone with contributor access can merge it.

Thanks for all the help and comments!

**Comment by daniel-beck** [comment]
@gdelory My apologies for not considering the following earlier, but:

There are different kinds of color blindness, and what's visible to one person might be difficult to see for another. While it's probably a good idea to make this configurable, it seems more useful to me to instead define sensible global defaults (it seems to be currently missing), and allow users to define their own preferences, similar to what Blue Balls Plugin does (although I wouldn't be surprised if its implementation isn't great).

I'm not maintaining or even using this plugin, just a thought.

**Comment by gdelory** [comment]
Hi Daniel, I agree with you that colorblindness is a user-dependant issue and choosing one color for everyone might not be the best option. However even for non-colorblind people the orange is really hard to read to it's still a nice thing to have. I would still like to go with this PR for now because I won't have time right away to implement an option for users to choose their own colors, but this is definitely something I'll try to do later this year.

I checked the Green (I admit Blue must have been a tempting name) Ball Plugin and they are using a User Property which could be a nice solution as well for this, but I'll have to find a colorpicker because we can't ask normal user to enter hex color code :)

If this PR get accepted before I have time to start working on it I'll open a new one with the user-defined colors.

**Comment by daniel-beck** [comment]
@gdelory 

> However even for non-colorblind people the orange is really hard to read to it's still a nice thing to have.

Agree. Which is why I propose to switch to sensible, works-for-most, non-configurable global defaults, and leave customization to per-user options.

**Comment by gdelory** [comment]
Haha true. I started with configurable colors but haven't even thought about just changing the orange to something better, which is a one line change, silly me :)
I'll submit a new PR with just this change this week. Hopefully it will be approved quickly. Then I'll try to find some time to implement the per-user options.
I'll probably go with "#db740" which is easier to read than "orange", let me know if any of you thinks it's hard to read as well.

**Comment by daniel-beck** [comment]
> Hopefully it will be approved quickly.

Seems like @recena was active most recently, so pinging him.

I wouldn't be comfortable acting here, as mentioned above:

> I'm not maintaining or even using this plugin

**Comment by hypery2k** [comment]
if you're still interested in this feature, please open a new PR against develop

**Pull request closed without merging**